src/base/bittorrent/tracker.cpp

```diff
@@ -32,235 +33,419 @@
+        const int num = numWantIter->toInt();
+        if (num < 0)
+            throw TrackerError("Invalid numwant parameter");
+        announceReq.numwant = num;
     }
 
-    // 6. no_peer_id (extension)
-    announceReq.noPeerId = false;
-    if (queryParams.contains("no_peer_id"))
-        announceReq.noPeerId = true;
+    // 5. no_peer_id
+    // non-formal extension
+    announceReq.noPeerId = (queryParams.value(ANNOUNCE_REQUEST_NO_PEER_ID) == "1");
+
+    // 6. left
+    announceReq.peer.isSeeder = (queryParams.value(ANNOUNCE_REQUEST_LEFT) == "0");
 
-    // 7. TODO: support "compact" extension
+    // 7. compact
+    announceReq.compact = (queryParams.value(ANNOUNCE_REQUEST_COMPACT) != "0");
 
-    // Done parsing, now let's reply
-    if (announceReq.event == "stopped") {
+    // 8. event
+    announceReq.event = queryParams.value(ANNOUNCE_REQUEST_EVENT);
+
+    if (announceReq.event.isEmpty()
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_EMPTY)
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_COMPLETED)
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_STARTED)
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_PAUSED)) {
+        // [BEP-21] Extension for partial seeds (partial support)
+        registerPeer(announceReq);
+        prepareAnnounceResponse(announceReq);
+    }
+    else if (announceReq.event == ANNOUNCE_REQUEST_EVENT_STOPPED) {
         unregisterPeer(announceReq);
     }
     else {
-        registerPeer(announceReq);
-        replyWithPeerList(announceReq);
+        throw TrackerError("Invalid event parameter");
     }
 }
 
 void Tracker::registerPeer(const TrackerAnnounceRequest &announceReq)
 {
-    if (announceReq.peer.port == 0) return;
-
     if (!m_torrents.contains(announceReq.infoHash)) {
-        // Unknown torrent
-        if (m_torrents.size() == MAX_TORRENTS) {
-            // Reached max size, remove a random torrent
+        // Reached max size, remove a random torrent
+        if (m_torrents.size() >= MAX_TORRENTS)
             m_torrents.erase(m_torrents.begin());
-        }
     }
 
-    // Register the user
-    PeerList &peers = m_torrents[announceReq.infoHash];
-    if (!peers.contains(announceReq.peer.uid())) {
-        // Unknown peer
-        if (peers.size() == MAX_PEERS_PER_TORRENT) {
-            // Too many peers, remove a random one
-            peers.erase(peers.begin());
-        }
+    TorrentStats &torrentStats = m_torrents[announceReq.infoHash];
+    QSet<Peer> &peers = torrentStats.peers;
+
+    // Too many peers, remove a random one
+    if (peers.size() >= MAX_PEERS_PER_TORRENT)
+        torrentStats.removePeer(*peers.begin());
+
+    const auto peerIter = peers.find(announceReq.peer);
+    if (peerIter == peers.end()) {
+        // is new peer
+        torrentStats.addPeer(announceReq.peer);
+    }
+    else {
+        // is existing peer, replace it
+        torrentStats.removePeer(*peerIter);
+        torrentStats.addPeer(announceReq.peer);
```

Comment:
I would have `setPeer` instead of `addPeer` so that it adds peer if there's no such one or update it if it is already registered.

Reply:
The problem is you can't update existing entries in-place, QSet doesn't support that.
See: https://doc.qt.io/qt-5/qset-iterator.html#operator-2a, there is no other non-const deference operator.

Reply:
>The problem is you can't update existing entries in-place, QSet doesn't support that.

It's not the problem. You can implement it via remove-then-insert. But TorrentStats will have easier interface.

Reply:
Changed.